src/components/cursor-controller.js

```diff
@@ -3,39 +3,113 @@ const TARGET_TYPE_INTERACTABLE = 2;
     this.data.cursor.addEventListener("loaded", this._handleCursorLoaded);
+
+    // raycaster state
+    this.targets = [];
+    this.intersection = null;
+    this.raycaster = new THREE.Raycaster();
+    this.setDirty = this.setDirty.bind(this);
+    this.dirty = true;
+  },
+
+  update: function() {
+    this.raycaster.far = this.data.far;
+    this.raycaster.near = this.data.near;
+    this.setDirty();
   },
 
+  play: function() {
+    this.observer = new MutationObserver(this.setDirty);
+    this.observer.observe(this.el.sceneEl, { childList: true, attributes: true, subtree: true });
+    this.el.sceneEl.addEventListener("object3dset", this.setDirty);
+    this.el.sceneEl.addEventListener("object3dremove", this.setDirty);
+  },
+
+  pause: function() {
+    this.observer.disconnect();
+    this.el.sceneEl.removeEventListener("object3dset", this.setDirty);
+    this.el.sceneEl.removeEventListener("object3dremove", this.setDirty);
+  },
+
+  setDirty: function() {
+    this.dirty = true;
+  },
+
+  populateEntities: function(selector, target) {
+    target.length = 0;
+    const els = this.data.objects ? this.el.sceneEl.querySelectorAll(this.data.objects) : this.el.sceneEl.children;
+    for (let i = 0; i < els.length; i++) {
+      if (els[i].object3D) {
+        target.push(els[i].object3D);
+      }
+    }
+  },
+
+  emitIntersectionEvents: function(prevIntersection, currIntersection) {
+    // if we are now intersecting something, and previously we were intersecting nothing or something else
+    if (currIntersection && (!prevIntersection || currIntersection.object.el !== prevIntersection.object.el)) {
+      this.data.cursor.emit("raycaster-intersection", { el: currIntersection.object.el });
+    }
+    // if we were intersecting something, but now we are intersecting nothing or something else
+    if (prevIntersection && (!currIntersection || currIntersection.object.el !== prevIntersection.object.el)) {
+      this.data.cursor.emit("raycaster-intersection-cleared", { el: prevIntersection.object.el });
+    }
+  },
+
+  performRaycast: (function() {
+    const rayObjectRotation = new THREE.Quaternion();
+    const rawIntersections = [];
+    return function performRaycast(targets) {
+      if (this.data.rayObject) {
+        const rayObject = this.data.rayObject.object3D;
+        rayObject.updateMatrixWorld();
+        rayObjectRotation.setFromRotationMatrix(rayObject.matrixWorld);
+        this.raycaster.ray.origin.setFromMatrixPosition(rayObject.matrixWorld);
+        this.raycaster.ray.direction.set(0, 0, -1).applyQuaternion(rayObjectRotation);
+      } else {
+        this.raycaster.setFromCamera(this.mousePos, this.data.camera.components.camera.camera); // camera
+      }
+      const prevIntersection = this.intersection;
+      rawIntersections.length = 0;
+      this.raycaster.intersectObjects(targets, true, rawIntersections);
+      this.intersection = rawIntersections.find(x => x.object.el);
```

Comment:
This is surprising to me. If the targets are `object3D`, then should this be `x.el`?